RecoLocalCalo/HcalRecAlgos/src/MahiGPU.cu

```diff
@@ -0,0 +1,1432 @@
+            ipulse, offset);
+#endif
+
+        // in order not to build matrcies  for pulseP/M
+        auto const ampl2 = resultAmplitude * resultAmplitude;
+        #pragma unroll
+        for (int counter=0; counter<NSAMPLES2; counter++) {
+            int const i = counter / NSAMPLES;
+            int const j = counter % NSAMPLES;
+            float const valueP_i = pulseMatrixP(i, ipulseReal);
+            float const valueP_j = pulseMatrixP(j, ipulseReal);
+            float const value_i = pulseMatrix(i, ipulse);
+            float const value_j = pulseMatrix(j, ipulse);
+            float const valueM_i = pulseMatrixM(i, ipulseReal);
+            float const valueM_j = pulseMatrixM(j, ipulseReal);
+            auto const covValue = 0.5 * (
+                (valueP_i - value_i) * (valueP_j - value_j) +
+                (valueM_i - value_i) * (valueM_j - value_j));
+            covarianceMatrix(i, j) += ampl2 * covValue;
+
+#ifdef HCAL_MAHI_GPUDEBUG
+            printf("%f ", covValue);
+            if ((counter+1) % NSAMPLES == 0)
+                printf("\n");
+#endif
+        }
+    }
+}
+
+template<int NSAMPLES, int NPULSES>
+__global__
+void kernel_minimize(
+        float* outputEnergy,
+        float* outputChi2,
+        float const* inputAmplitudes,
+        float* pulseMatrices,
+        float* pulseMatricesM, // should be const but Eigen complains
+        float* pulseMatricesP, // hsould be const but eigen complains
+        int const* pulseOffsetValues,
+        float const* noiseTerms,
+        int8_t const* soiSamples,
+        float const* pedestalWidths,
+        float const* effectivePedestalWidths,
+        bool const useEffectivePedestals,
+        uint32_t const* idsf01HE,
+        uint32_t const* idsf5HB,
+        float const* gainValues,
+        float const* respCorrectionValues,
+        uint32_t const nchannelsf01HE,
+        uint32_t const nchannelsTotal,
+        uint32_t const offsetForHashes,
+        int const maxDepthHB,
+        int const maxDepthHE,
+        int const maxPhiHE,
+        int const firstHBRing,
+        int const lastHBRing,
+        int const firstHERing,
+        int const lastHERing,
+        int const nEtaHB,
+        int const nEtaHE) {
+    // indices
+    auto const gch = threadIdx.x + blockIdx.x * blockDim.x;
+    if (gch >= nchannelsTotal) return;
+
+    // if chi2 is set to -9999 do not run minimization
+    if (outputChi2[gch] == -9999.f)
+        return;
+
+    // conditions for pedestal widths
+    auto const id = gch >= nchannelsf01HE
+        ? idsf5HB[gch - nchannelsf01HE]
+        : idsf01HE[gch];
+    auto const did = DetId{id};
+    auto const hashedId = did.subdetId() == HcalBarrel
+        ? did2linearIndexHB(id, maxDepthHB, firstHBRing, lastHBRing, nEtaHB)
+        : did2linearIndexHE(id, maxDepthHE, maxPhiHE, firstHERing, lastHERing, nEtaHE)
+            + offsetForHashes;
+
+    auto const* pedestalWidthsForChannel = useEffectivePedestals && gch < nchannelsf01HE
```

Comment:
why here you are still accessing the objects with hashedId ? should be better to calculate those pedestal when preparing the other inputs as amplitudes  and noise vector ? and store in memory ?

Reply:
no, as it will require additional transfers per event. u transfer conditions once and u access conditions based on the hash, no point to do anything further.